### PR TITLE
WT-10967 Remove the gcc.h comment and replace it with documentation pages.

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [inline src/docs/arch-concurrency.dox:15]: Comment from Chenhao: Can we explain why we need to define the frequently used concurrent variables as volatile?

Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:15]: No I cannot, personally I don't think we should define variables as volatile and instead use read_once write_once. But that would be too large a change and so I put this sentence in to manage that.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [inline src/docs/arch-concurrency.dox:39]: We explain too much about locks here, we should leave it to the reader to know what a lock is and simply discuss WiredTiger implementation and usage.

File: src/docs/arch-index.dox
Comment (luke-pearson) [inline src/docs/arch-index.dox:284]: This will be paired with an `arch-usage-pattern` doc, coming soon.™

File: src/docs/arch-concurrency.dox
Comment (quchenhao) [inline src/docs/arch-concurrency.dox:70]: Better to point out relaxed consistency defined in the standard memory model

File: src/docs/arch-concurrency.dox
Comment (quchenhao) [inline src/docs/arch-concurrency.dox:74]: also compiling reordering

File: src/docs/arch-concurrency.dox
Comment (quchenhao) [inline src/docs/arch-concurrency.dox:100]: What is the optimization?

File: src/docs/arch-concurrency.dox
Comment (ajmorton) [inline src/docs/arch-concurrency.dox:70]:
```suggestion
transition atomic load and store macros have been added to WiredTiger. To prevent them from introducing a performance regression they utilize relaxed consistency.
```

File: src/docs/arch-concurrency.dox
Comment (ajmorton) [inline src/docs/arch-concurrency.dox:77]: I'd simplify to
```suggestion
seeing incorrect state.
```
in case "impossible values" raises questions

File: src/docs/tools/doxfilter.py
Comment (ajmorton) [inline src/docs/tools/doxfilter.py:79]: Why was this change needed?

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [inline src/docs/arch-concurrency.dox:67]: I think the documentation should concern itself with the code "as it is" at the time this document is added/updated.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [inline src/docs/arch-concurrency.dox:43]: I think this should be considered assumed reader knowledge. Multi-threaded programming is relatively commonplace.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [inline src/docs/arch-concurrency.dox:45]: Talking about the generation management design is valuable, but I find it incongruous with the other low-level descriptions in this document. Perhaps a later section or separate document with other data tracking mechanisms would flow better.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [inline src/docs/arch-concurrency.dox:90]: I suggest each WT barrier type should be enumerate into its own paragraph describing the purpose. I don't think a lot of detail is required.
Also potentially a link to the Preshing resource for further reference.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [inline src/docs/arch-concurrency.dox:92]: I thought this has been deprecated, and no longer needs to be described.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [inline src/docs/arch-concurrency.dox:28]: I think any discussion of WT portable synchronization interface should be in a paragraph/section of it's own and only need mention that it exist, and is intended to give equivalent semantics on all supported platforms.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [inline src/docs/arch-concurrency.dox:35]: Two of the 3 are not true spinlocks. @ajmorton has already mentioned the relevant ticket WT-12960.

File: src/docs/arch-concurrency.dox
Comment (jeremythorp) [reply src/docs/arch-concurrency.dox:9]: I agree with @ajmorton and @marcbutler on this one.

**Comment by jeremythorp** [comment]
@luke-pearson , can I make a broader suggestion for efficiently progressing this PR: as the main change is to create a new section of technical documentation, working on and reviewing the new text in a Google Doc would be much easier than using PR style reviews. Then, once we're all agreed on the text in Google Doc it would be simply copy-pasted with the right formatting into the doc source ready for a quick PR review.

**Comment by luke-pearson** [comment]
> @luke-pearson , can I make a broader suggestion for efficiently progressing this PR: as the main change is to create a new section of technical documentation, working on and reviewing the new text in a Google Doc would be much easier than using PR style reviews. Then, once we're all agreed on the text in Google Doc it would be simply copy-pasted with the right formatting into the doc source ready for a quick PR review.

You can but I'm going to decline. This content was largely adapted from prior code comments or previously reviewed PM-3220 docs.

File: src/docs/tools/doxfilter.py
Comment (luke-pearson) [reply src/docs/tools/doxfilter.py:79]: It stops some boilerplate from being added to the doc. The original architecture guide pages were added to match up with the aritecture diagram here: https://source.wiredtiger.com/develop/arch-index.html.

As this content does seem to fit into the architecture guide but not that diagram I spoke to Don and he showed me how to disable the boilerplate.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:43]: This speaks to a specific WiredTiger methodology, while I don't agree with our usage I was simply documenting it. As otherwise we have no documentation for this oddity.

File: src/docs/portability.dox
Comment (luke-pearson) [reply src/docs/portability.dox]: [7353ef8](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/7353ef88edcfd2c352df82dd082ad089faab6399)

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: [7353ef8](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/7353ef88edcfd2c352df82dd082ad089faab6399)

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:17]: Removed majority of it in: 9599c06fc. The section on lock usage will persist until this https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1612707512 comment chain is resolved.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:76]: This section exists to explain and define the usage of the phrase publish in the codebase. Which I think is valuable as previously a long discussions was held to agree on the definition. If we do not codify this definition in the documentation we risk losing that discussion and will eventually do it over again.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: Removed per this conversation: https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1612519348

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: Removed per this conversation: https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1612519348

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:29]: As we do not currently have that document / page, would it be acceptable to move it at a later time? We will create an arch-usage patterns page but that talks about higher level data stuctures and algorithms within WiredTiger, which generations likely doesn't fall into either.

As this page has the heading "WiredTiger concurrency management" I think generations belongs here.

In theory the generation system is interesting enough and could have enough detail to have it's own page in the docs. However writing that page is not within the scope of PM-3221 and would fall under a separate documentation ticket.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: Nuked the spin lock reference in b3cdd6cab

**Comment by luke-pearson** [comment]
I've removed spin locks from the doc in b3cdd6cab.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:50]: To be fair the "as it is" state of the code is that WiredTiger is in the middle of transitioning to TSan, which matches the written documentation. That said I understand your point to be that documentation shouldn't really describe a transitional state.

This means I can either speak to pre or post TSan. I will opt to speak to the pre TSan form of the codebase which means to delete this paragraph in it's entirety. Is that acceptable @marcbutler?

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: 993bc3174

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: [7c8a76d](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/7c8a76d273ceb9d3513e47229079367559a7b4c8)

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: Removed in [7c8a76d](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/7c8a76d273ceb9d3513e47229079367559a7b4c8). But see this blog post https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/enabling-rcpc-in-gcc-and-llvm

**Comment by jeremythorp** [comment]
> > @luke-pearson , can I make a broader suggestion for efficiently progressing this PR: as the main change is to create a new section of technical documentation, working on and reviewing the new text in a Google Doc would be much easier than using PR style reviews. Then, once we're all agreed on the text in Google Doc it would be simply copy-pasted with the right formatting into the doc source ready for a quick PR review.
> 
> You can but I'm going to decline. This content was largely adapted from prior code comments or previously reviewed PM-3220 docs.

Hi @luke-pearson , the goal is for us all to reach agreement on the revised text. If we can do that quickly and efficiently using the current PR comment approach then that's great, but if not we should switch temporarily to a Google Doc to help speed up any complex discussion.

File: src/docs/arch-concurrency.dox
Comment (ajmorton) [reply src/docs/arch-concurrency.dox:26]: +1 
Having variables that are only sometimes protected by locks was enough of a surprise that I'd like to document it.

File: src/docs/arch-concurrency.dox
Comment (ajmorton) [inline src/docs/arch-concurrency.dox:101]: If you take the suggestions above we can reword this section to reference it, and talk about how marked accesses achieve the semantics we wanted from ACQUIRE_READ_WITH_BARRIER

File: src/docs/arch-concurrency.dox
Comment (ajmorton) [inline src/docs/arch-concurrency.dox:109]: "Reordered across it is" a bit vague. Please reword here and in the RELEASE_WRITE section.

File: src/docs/arch-concurrency.dox
Comment (ajmorton) [inline src/docs/arch-concurrency.dox:89]: Can we make this `full-bar` instead of adding `ful` to string.ok?

**Comment by luke-pearson** [comment]
> The goal is for us all to reach agreement on the revised text. If we can do that quickly and efficiently using the current PR comment approach then that's great.

I feel like we are making good progress here so I'll keep it here.

File: src/docs/arch-concurrency.dox
Comment (quchenhao) [inline src/docs/arch-concurrency.dox:84]: I feel we should further downplay this. May be we should remove this section.

Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:84]: This discussion is tracked here: https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1612719663.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: From Chenhao: "I feel we should further downplay this. May be we should remove this section."

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [reply src/docs/arch-concurrency.dox:50]: Is this expected to be a sustained transitional state: or is there a complete timeline for the TSan work? If there is a completion date: does it make sense to postpone updating this documentation until that work has been complete?

Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:50]: I think we can just leave it out and create a ticket to add this paragraph at the completion of that work. Thoughts?

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:3]: I prefer shared memory here, as the second sentence makes more sense. " Many mechanisms are defined in WiredTiger to safely access shared memory locations"

Comment (marcbutler) [reply src/docs/arch-concurrency.dox:3]: My intention was to communicate the term "shared memory" should not be used in this paragraph/document. For threads all memory is shared, the alternative is explicitly named: TLS (Thread Local Storage).

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [reply src/docs/arch-concurrency.dox:26]: IIRC we access without lock protection when it is thought that multi-threaded access is not possible due to the state of the system. I think explaining this reasoning is appropriate.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [reply src/docs/arch-concurrency.dox:50]: My global observation is that we should defer documenting anything that is considered short-term unstable/transitional.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:26]: Unfortunately that is not the case, the worst offender is the txn_global structure which has values read from it all over the place outside the txn_global rwlock. Two examples:

- https://github.com/wiredtiger/wiredtiger/blob/7facd2810f02f65be3a5530974aa3fc7b645daae/src/txn/txn_timestamp.c#L651
- https://github.com/wiredtiger/wiredtiger/blob/7facd2810f02f65be3a5530974aa3fc7b645daae/src/include/txn_inline.h#L660

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: This was deleted.

Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: It was removed.

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: Changed the phrasing in dd47e7333

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: e40148170

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: e40148170

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: e40148170

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox:103]: 6cb86ad91

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: 6cb86ad91

File: src/docs/spell.ok
Comment (ajmorton) [inline src/docs/spell.ok:391]:
```suggestion
```

File: src/docs/spell.ok
Comment (ajmorton) [inline src/docs/spell.ok:616]: Sorry, missed this in the prior convo about `ful`. Can we use `release` in the labels and not whitelist this?

**Review by ajmorton — APPROVED** [review]
LGTM pending comments from other reviewers

File: src/docs/spell.ok
Comment (luke-pearson) [reply src/docs/spell.ok:616]: acq-rel are fairly standard shortenings does it really matter that we've added a spelling to this arbitrary list?

Comment (ajmorton) [reply src/docs/spell.ok:616]: Yeah fair. Happy to leave this as is.

File: src/docs/arch-concurrency.dox
Comment (marcbutler) [reply src/docs/arch-concurrency.dox:29]: 👍 Fair call.

**Review by marcbutler — APPROVED** [review]
LGTM
Appreciate all your efforts.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/docs/spell.ok
Comment (luke-pearson) [reply src/docs/spell.ok]: aaa8395850d49de764fcfa9d9368278df82cb28f

File: src/docs/arch-concurrency.dox
Comment (jeremythorp) [reply src/docs/arch-concurrency.dox:27]: I agree with @luke-pearson and @ajmorton that it's worth specifically mentioning that variables that are only sometimes protected is a good idea. Otherwise readers of the code might incorrectly assume they've found a bug, or they might misunderstood how the code works in practice.

File: src/docs/arch-concurrency.dox
Comment (jeremythorp) [inline src/docs/arch-concurrency.dox:31]:
```suggestion
Shared objects, including the page index, WT_REFs, etc. have their lifecycles managed using
```

File: src/docs/arch-concurrency.dox
Comment (jeremythorp) [inline src/docs/arch-concurrency.dox:117]:
```suggestion
with it. Ensuring the store happens after the given loads and stores. As with the acquire and
```

**Review by jeremythorp — APPROVED** [review]
With a couple of trivial tweaks this LGTM!
Thanks everyone for contributing to an effective review!

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: [3d71dfb](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/3d71dfb97c01ae20ef311d37ef3b0a5a24e7d735)

File: src/docs/arch-concurrency.dox
Comment (luke-pearson) [reply src/docs/arch-concurrency.dox]: [3d71dfb](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/3d71dfb97c01ae20ef311d37ef3b0a5a24e7d735)

**Pull request merged**
